### ISSUE TITLE
Performance for full-scan select concurrent with insert gets slow

### ISSUE THREAD

**Issue by mdcallag** [orioledb/orioledb#277]
During the load benchmark step for sysbench that loads 20M rows I did a few "select count(*) from sbtest1" while the table is being loaded. This table fits in the OS page cache, but not in the 1G OrioleDB buffer pool. The load does a commit per insert so it doesn't have a long-open transaction.

The query takes less than 1 second until the table gets ~2M rows. At that point the query takes a really long time and it has been 12 minutes so far. The load finished minutes ago, the next step after the load is creating indexes but ps shows "CREATE INDEX waiting" perhaps because the SELECT is still running.

Also, if I do a control-C in the shell running the query, I get a message about the cancel being sent but the psql client process doesn't stop.

Does Postgres have a check-for-interrupt macro so that backends can check to see if they should stop processing a query?

The thread stack for the Postgres backend handling the SELECT is ...
```
      1 __libc_pread64,pread,FileRead,OFileRead,read_buffer,get_buffer,o_buffers_rw,get_page_from_undo,load_first_historical_page,iterate_internal_page,btree_seq_scan_getnext_internal,btree_seq_scan_getnext,orioledb_getnextslot,table_scan_getnextslot,SeqNext,ExecProcNode,fetch_input_tuple,agg_retrieve_direct,ExecAgg,ExecProcNode,ExecutePlan,standard_ExecutorRun,orioledb_ExecutorRun_hook,PortalRunSelect,PortalRun,exec_simple_query,PostgresMain,BackendRun,BackendStartup,ServerLoop,PostmasterMain,main
```

**Comment by mdcallag** [comment]
10 hours later and the query is still running. The stack traces I get with PMP is the same as above.

**Comment by akorotkov** [comment]
@mdcallag, thank you! I'll recheck this.

**Comment by pashkinelfe** [comment]
@mdcallag what is the size of your row? Does it fit 2Kb per row for storage PLAIN or the values are TOASTed?

**Comment by mdcallag** [comment]
Create table [is here](https://github.com/mdcallag/mytools/blob/master/bench/ibench/iibench.py#L408-L424)

The size of entries in the data column is configurable at benchmark run time and I set it to be ~20 bytes. It will be less than 2000 bytes with my usage.

```
def create_table_postgres():
  conn = get_conn()
  cursor = conn.cursor()
  cursor.execute('drop table if exists %s' % FLAGS.table_name)

  if FLAGS.num_partitions > 0:
    part_sql = 'partition by range( transactionid )'
  else:
    part_sql = ''

  col_sql = 'transactionid bigserial primary key, '\
            'dateandtime timestamp without time zone, '\
            'cashregisterid int not null, '\
            'customerid int not null, '\
            'productid int not null, '\
            'price real not null, '\
            'data varchar(4000) '
```

**Comment by mdcallag** [comment]
> I've tried to reproduce in the following way: A table with tuple of ~40 Bytes:
> 
> ```
>  id    | project_id |             ts             |              filler
> ----------+------------+----------------------------+----------------------------------
>  11054737 |       6901 | 2023-10-04 12:45:46.050807 | 9e163c211184c8771db6517ce781b135
> ```
> 
> Inserted tuples with one process into a table. TPS was approximately constant ~1500. I've inserted 50M rows, and regularly (every 50s) was doing select count(*).

I shared the wrong DDL above, that was from the Insert Benchmark and I am using sysbench for now. The correct DDL is:
```
ib=# \d+ sbtest1
                                                           Table "public.sbtest1"
 Column |      Type      | Collation | Nullable |               Default               | Storage  | Compression | Stats target | Description
--------+----------------+-----------+----------+-------------------------------------+----------+-------------+--------------+-------------
 id     | integer        |           | not null | nextval('sbtest1_id_seq'::regclass) | plain    |             |              |
 k      | integer        |           | not null | 0                                   | plain    |             |              |
 c      | character(120) |           | not null | ''::bpchar                          | extended |             |              |
 pad    | character(60)  |           | not null | ''::bpchar                          | extended |             |              |
Indexes:
    "sbtest1_pkey" PRIMARY KEY, btree (id)
Access method: orioledb
```

My test server has 8 cores, 16G of RAM and this configuration right now. While below it has orioledb.main_buffers = 4GB, I reproduced the problem with that set to 1GB and 8GB.
```
huge_pages =    off                     # on, off, or try
max_parallel_workers_per_gather = 0     # taken from max_parallel_workers
max_parallel_workers = 0                # maximum number of max_worker_processes that
checkpoint_warning = 1800s               # 0 disables
wal_sync_method = fdatasync             # the default is the first option
wal_level='replica'                     # to log what replication needs
full_page_writes = on                   # recover from partial page writes
track_io_timing = on
synchronous_commit = off                # synchronization level;
#shared_buffers = 10GB
shared_buffers = 128MB
bgwriter_lru_maxpages =  2000           # max buffers written/round, 0 disables
bgwriter_delay = 100ms                  # 10-10000ms between rounds
wal_compression = off
max_wal_size = 40GB
min_wal_size = 20GB
log_autovacuum_min_duration=0
log_checkpoints = on

checkpoint_timeout=1800s
autovacuum_max_workers=4
autovacuum_naptime=10s
autovacuum_vacuum_cost_delay=2ms
autovacuum_vacuum_cost_limit=4000
autovacuum_vacuum_scale_factor=0.05
autovacuum_vacuum_insert_scale_factor=0.05
autovacuum_work_mem=256MB

# the size of shared memory, where hot data pages of OrioleDB tables are cached. This parameter is analog of the built-in shared_buffers GUC parameter. Default is 64 MB.
orioledb.main_buffers = 4GB

# shared memory size for metadata of block allocators for compressed tables. The default is 8 MB. We recommend increasing the value of this parameter to work with large compressed tables.
# orioledb.free_tree_buffers

# shared memory size of table metadata. The default value is 8 MB. We recommend increasing the value of this parameter to work with a large number of tables.
orioledb.catalog_buffers = 64MB

# the shared memory ring buffer size for older versions of rows and pages. The default is 1 MB.
orioledb.undo_buffers = 256MB

# the number of recovery workers row-level WAL based recovery. The default is 3. We recommend increasing the value of this parameter for the systems with a large number of CPU cores.
# orioledb.recovery_pool_size

# the size of shared memory for message queues related to recovery workers. The default is 8 MB.
orioledb.recovery_queue_size = 64MB

# the fraction of OrioleDB tables checkpoint time within the whole checkpoint time. The default is 0.5. We recommend setting this value to 1.0 if only OrioleDB tables are used.
orioledb.checkpoint_completion_ratio = 1.0

# the number background writer processes, which flushes dirty pages of OrioleDB tables in background. We recommend setting values greater than 1 for the systems with a large number of CPU cores. The default is 1.
orioledb.bgwriter_num_workers = 3
```

I am using sysbench and the command line is:
```
ts=20000000; /home/mdcallag/sysb/bin/sysbench --db-driver=pgsql --range-size=100 --table-size=${ts} --tables=1 --events=0 --time=600 /home/mdcallag/sysb/share/sysbench/oltp_point_select.lua prepare --pgsql-user=root --pgsql-password=pw --pgsql-host=127.0.0.1 --pgsql-db=ib --create-table-options="using orioledb"
```

While doing the load I run the select count(*) from sbtest1 query in a loop. The output below is an example, the queries are finishing in less than one second, but one the count passes 2M the next query never finishes.
```
  count
---------
 1899646
(1 row)


real    0m0.733s
user    0m0.001s
sys     0m0.001s
  count
---------
 1947672
(1 row)


real    0m0.661s
user    0m0.002s
sys     0m0.000s
  count
---------
 1993029
(1 row)


real    0m0.726s
user    0m0.002s
sys     0m0.000s
  count
---------
 2041053
(1 row)


real    0m0.895s
user    0m0.002s
sys     0m0.000s

---> and then the next run of the query just pins a CPU core at 100% and never finishes
```

**Comment by mdcallag** [comment]
Well, this hasn't been my best bug report.
* I used the OrioleDB Postgres fork at patches16_18
* I used the OrioleDB extension at bbfc8ab2ba6 from Sep 10 ("Stamp beta 2")
* rows should be ~200 bytes
* inserts are done from one connection
* OOM and the OOM killer hasn't been an issue here, but I have another bug for that, https://github.com/orioledb/orioledb/issues/276
* config is listed above, server has 8 cores and 16G of RAM

I repeated the test with orioledb.undo_buffers increased from 256MB to 1GB and 4GB and the problem repeats at the same point in time, after inserting ~2M rows.

I repeated the test with synchronous_commit=ON and the problem repeats at a similar point in time.

**Comment by mdcallag** [comment]
I have two builds - one that I trust and one that I don't trust. Right now I cannot reproduce this using the build that I trust. I am not certain about which build I was using when I could reproduce this. Debugging future problems would be slightly easier if git hashes were written to the Postgres logfile at startup.

With the build that I trust, I cannot reproduce this with the query running in a loop regardless of how long I sleep in between queries. I tried from 0 to 10 seconds of sleep after a query.

So I will close this, and revisit the other bugs I filed.
